File: kotlin-back/src/main/kotlin/com/example/kotlinback/global/generator/IdGenerator.java
package com.example.kotlinback.global.generator;

import org.hibernate.HibernateException;
import org.hibernate.MappingException;
import org.hibernate.engine.spi.SharedSessionContractImplementor;
import org.hibernate.id.Configurable;
import org.hibernate.id.IdentifierGenerator;
import org.hibernate.internal.util.config.ConfigurationHelper;
import org.hibernate.service.ServiceRegistry;
import org.hibernate.type.Type;

import java.io.Serializable;
import java.time.LocalDate;
import java.util.Properties;
import java.util.UUID;

public class IdGenerator implements IdentifierGenerator {
    @Override
    public Serializable generate(SharedSessionContractImplementor session,
                                 Object object) throws HibernateException {
        return "%s%s".formatted(
                LocalDate.now().toString().replace("-", ""),
                UUID.randomUUID().toString()
        );
    }
}

File: kotlin-back/src/main/kotlin/com/example/kotlinback/common/generator/IdGenerator.kt
package com.example.kotlinback.common.generator

import org.hibernate.HibernateException
import org.hibernate.MappingException
import org.hibernate.engine.spi.SharedSessionContractImplementor
import org.hibernate.id.Configurable
import org.hibernate.id.IdentifierGenerator
import org.hibernate.internal.util.config.ConfigurationHelper
import org.hibernate.service.ServiceRegistry
import org.hibernate.type.Type
import java.io.Serializable
import java.time.LocalDate
import java.util.*

class IdGenerator : IdentifierGenerator, Configurable {
    private var entityType: String? = null

    @Throws(HibernateException::class)
    override fun generate(
        session: SharedSessionContractImplementor,
        `object`: Any
    ): Serializable {
        return "%s%s%s".formatted(
            LocalDate.now().toString().replace("-", ""),
            entityType,
            UUID.randomUUID().toString()
        )
    }

    @Throws(MappingException::class)
    override fun configure(
        type: Type,
        params: Properties,
        serviceRegistry: ServiceRegistry
    ) {
        entityType = ConfigurationHelper.getString(ENTITY_TYPE, params)
    }

    companion object {
        const val ENTITY_TYPE = "entityType"
    }
}
